### ISSUE TITLE
Parallel text vectorization 

### ISSUE THREAD

**Issue by rth** [FreeDiscovery/FreeDiscovery#152]
Text vectorization is currently one of the slowest steps in the processing. There were proposals to parallelize scikit-learn's text vectorizers e.g. in https://github.com/scikit-learn/scikit-learn/pull/2923 but they didn't end up getting merged into scikit-learn. 

Another approach, [as suggested](https://github.com/scikit-learn/scikit-learn/pull/6473#issuecomment-308602876) by @jnothman would be to apply the existing vectorizers on chunks of the corpus, then merge the results.

For `CountVectorizer` and `TfidfVectorizer` the processing could be, as follows,
   1. Apply the vectorizers in parallel on chunked text corpus (with `use_idf=False`) , and compute the document-term matrices `X_k` (CSR sparse array) and vocabularies (`vocabulary_k`) for each chunk
   2. Merge all `vocabulary_k` into a common vocabulary, and compute the numeric offsets (`vocabulary_offset_k`) between the individual vocabularies and the common vocabulary
   3. For row `j` of each chunk: `X_k[j].indices = X_k[j].indices + vocabulary_offset_k` (need to see how to actually do this), also make sure it's still a valid CSR array
   4. Concatenate the results with `scipy.sparse.vstack`
   5. (optional) Apply the `TfidfTransformer` if necessary 

One could do this either by subclassing scikit-learn's vectorizers or possibly with a single class decorator that could work for all vectorizers...

For the `HashingVectorizer` steps 2, 3 are not necessary since the hashed vocabularies are common by definition. This is already used in FreeDiscovery [here](https://github.com/FreeDiscovery/FreeDiscovery/blob/master/freediscovery/text.py#L459) but this should be refactored to expose an independent Python class with a scikit-learn compatible API..  cc @psinger

**Comment by jnothman** [comment]
One problem with merging learnt CountVectorizers is that you cannot use
min_df, max_df, max_features, but may want to apply them after the fact. Of
course, you can estimate your stop-words as would be removed by max_df on a
sub-sample.

I don't think your vocabulary offsets make sense. If you don't mind the
sort time, you can use

```
all_feature_names, mappings = np.unique(sum((vec.get_feature_names() for
vec in vecs), []), return_inverse=True)
```

**Comment by rth** [comment]
Thanks for the feedback @jnothman ! Yes, correctly handling of `min_df`, `max_features` etc in this case would require additional effort. About the vocabulary offsets, you are probably right: prototyped it a bit quickly; I will look into it. Thanks!
